### PR TITLE
fix(http): REST interface sends HTTP error replies on runtime errors

### LINKED ISSUE #3863: The REST interface closes a connection on implicit casting error
### Describe the bug

This test:
```java
    @Test
    public void testBadImplicitStrToLongCast() throws Exception {
        getSimpleTester().run(engine -> {
            testHttpClient.assertGet("{\"ddl\":\"OK\"}", "create table tab (value int, when long, ts timestamp) timestamp(ts) partition by day wal;");
            testHttpClient.assertGet("{\"ddl\":\"OK\"}", "insert into tab values(1, now(), now());"); // it should not be DDL:OK. change me when https://github.com/questdb/questdb/issues/3858 is fixed
            drainWalQueue(engine);

            testHttpClient.assertGet(
                    "{\"query\":\"select * from tab where when = '2023-10-17';\",\"error\":\"inconvertible value: `2023-10-17` [STRING -> LONG]\",\"position\":31}",
                    "select * from tab where when = '2023-10-17';"
            );
        });
    }
```
It fails to parse `2023-10-17` as a long (which is expected), but the error is not propagated to a client - it's only logged and then server closes the connection. It probably treats the implicit casting exception as a critical exception? 

I believe @puzpuzpuz said the PGWire interface behaves better, but it should be double-checked - in the case I misunderstood him. 

### To reproduce

_No response_

### Expected Behavior

_No response_

### Environment

```markdown
- **QuestDB version**:
- **OS**:
- **Browser**:
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jerrinot**
fixes #3863

**Old behaviour**
When a SELECT/PSEUDO_SELECT SQL is successfully compiled then QuestDB optimistically sends HTTP 200. Later, if there is an error while iterating over the record cursor then it would simply close the connection. Because it cannot take back the already sent HTTP 200 header. As a result, the client is left without any feedback and one has to explore server logs to learn what happened. A runtime error can happen e.g. due to an implicit casting error.

**New behaviour**
It fetches the first record and only if that's successful it sends HTTP 200. This gives us a chance to send an error HTTP code if we fail at runtime when accessing the first record. 

The fix is only partial. It's still possible for error to happen later. Due to lazy evaluation, etc. In this case, it behaves as before: close the connection. I'm afraid a general fix would require full result set materialization, which is a no-go. I believe many runtime errors happen when accessing the first record so this fix should work for many cases.

### CODE REVIEW

**Comment by jerrinot** [comment]
just trying what tests fail

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 14 / 14 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessorState.java|14|14|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 20 / 21 (95.24%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessorState.java|18|19|94.74%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessor.java|2|2|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 21 / 22 (95.45%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessorState.java|18|19|94.74%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpHeaderParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessor.java|2|2|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 22 / 23 (95.65%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessorState.java|18|19|94.74%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpHeaderParser.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessor.java|2|2|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 22 / 23 (95.65%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessorState.java|18|19|94.74%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpHeaderParser.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessor.java|2|2|100.00%|

File: core/src/main/java/io/questdb/cutlass/http/processors/JsonQueryProcessorState.java [hunk 1]
```diff
@@ -51,14 +51,15 @@
 
 public class JsonQueryProcessorState implements Mutable, Closeable {
     public static final String HIDDEN = "hidden";
+    static final int QUERY_SETUP_FIRST_RECORD = 0;
```
Comment (ideoma) [inline core/src/main/java/io/questdb/cutlass/http/processors/JsonQueryProcessorState.java:54]: nit: this file is not formatted, please reformat.

Comment (jerrinot) [reply core/src/main/java/io/questdb/cutlass/http/processors/JsonQueryProcessorState.java]: done

**Comment by jerrinot** [comment]
@ideoma thanks for reviewing. the test should be stabilized.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 38 / 39 (97.44%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessorState.java|32|33|96.97%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpHeaderParser.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/processors/JsonQueryProcessor.java|4|4|100.00%|

**Review by bluestreak01 — APPROVED** [review]

**Pull request merged**